modules/brushingandlinking/src/brushingandlinkingmanager.cpp

```diff
@@ -31,78 +31,422 @@
 #include <modules/brushingandlinking/processors/brushingandlinkingprocessor.h>
 #include <modules/brushingandlinking/ports/brushingandlinkingports.h>
 
+#include <inviwo/core/io/serialization/serializer.h>
+#include <inviwo/core/io/serialization/deserializer.h>
+#include <inviwo/core/util/exception.h>
+#include <inviwo/core/util/stdextensions.h>
+#include <inviwo/core/util/zip.h>
+#include <inviwo/core/util/assertion.h>
+
+#include <fmt/format.h>
+
 namespace inviwo {
 
-BrushingAndLinkingManager::BrushingAndLinkingManager(Processor* p,
+BrushingAndLinkingManager::BrushingAndLinkingManager(BrushingAndLinkingInport* inport,
                                                      InvalidationLevel validationLevel)
-    : owner_{p}, invalidationLevel_{validationLevel} {
-    auto outPorts = p->getOutports();
-    for (auto& op : outPorts) {
-        if (dynamic_cast<BrushingAndLinkingOutport*>(op)) {
-            op->onDisconnect([=]() { filtered_.update(); });
+    : owner_(inport), invalidationLevel_(validationLevel) {
+
+    inport->onConnect([&]() {
+        setParent(&static_cast<BrushingAndLinkingOutport*>(
+                       std::get<BrushingAndLinkingInport*>(owner_)->getConnectedOutport())
+                       ->getManager());
+    });
+    inport->onDisconnect([&]() { setParent(nullptr); });
+}
+
+BrushingAndLinkingManager::BrushingAndLinkingManager(BrushingAndLinkingOutport* outport,
+                                                     InvalidationLevel validationLevel)
+    : owner_(outport), invalidationLevel_(validationLevel) {
+
+    outport->onDisconnect([&]() { updatePortConnections(); });
+}
+
+BrushingAndLinkingManager::~BrushingAndLinkingManager() = default;
+
+void BrushingAndLinkingManager::brush(BrushingAction action, BrushingTarget target,
+                                      const BitSet& indices, std::string_view source) {
+    if ((action == BrushingAction::Filter) && source.empty()) {
+        throw Exception("BrushingAction::Filter requires a source", IVW_CONTEXT);
+    }
+
+    const int actionIdx = getActionIndex(action);
+
+    std::visit(util::overloaded{[&](BitSetTargets& map) { map[target] = indices; },
+                                [&](IndexListTargets& map) {
+                                    auto it = map.try_emplace(target, IndexList());
+                                    it.first->second.set(source, indices);
+                                }},
+               selections_[actionIdx]);
+
+    propagate(action, target);
+}
+
+bool BrushingAndLinkingManager::isModified() const { return !modifiedActions_.empty(); }
+
+BrushingModifications BrushingAndLinkingManager::modifiedActions() const {
+    return modifiedActions_;
+}
+
+bool BrushingAndLinkingManager::modifiedFiltering() const {
+    return bool(modifiedActions_ & BrushingModification::Filtered);
```

Comment:
```suggestion
    return modifiedActions_.contains(BrushingModification::Filtered);
```
I think that is much easier to read